tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/signal_kokkos_impl_called.hpp

```diff
@@ -0,0 +1,69 @@
+ /*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 2.0
+//              Copyright (2019) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_UTILS_HPP_
+#define LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_UTILS_HPP_
+
+#include <string_view>
+
+// Note: dependencies required only for compiling tests
+#if defined(KOKKOS_STDBLAS_ENABLE_TESTS)
+#include <iostream>
+#endif
+
+namespace KokkosKernelsSTD {
+namespace Impl {
+
+// For testing: signal that Kokkos implementation was called
+inline void signal_kokkos_impl_called(std::string_view functionName)
```

Comment:
Since this print is only enabled in debug mode anyway, what do we gain from doing this?

Reply:
What you would gain from doing this:

1. Not repeating a macro-guarded include of `<iostream>` at the top of every kokkos-kernels-based header file
2. Being able to change your tracing code with a relink, without needing to recompile everything

Reply:
Ok, let's do it. I haven't caught your intent behind `extern` at first.  
Please see https://github.com/kokkos/stdBLAS/pull/170/commits/d2c0ed546d8a1ad3e96696780da8e82f51016fa9:
1. In the client function we have single clean call - without `#ifdef`.
2. In `signal_kokkos_impl_called.hpp` header we have `extern` for testing and empty/NOP version otherwise (so we don't need any linkage outside of testing).
3. Linked implementation is on the testing side - completely separated from the main code now.

Can the function live in `gtest_main_kokkos.cpp` or would you prefer separate `signal_kokkos_impl_called.cpp` file for it ?

Reply:
Thanks!  Your approach is a bit more complicated than I had imagined, but not in a bad way.

> Can the function live in `gtest_main_kokkos.cpp` or would you prefer separate `signal_kokkos_impl_called.cpp` file for it ?

If the function lives in `gtest_main_kokkos.cpp`, then you shouldn't have to worry about `__declspec(dllexport)` for Windows builds.